src/Algebra/Graph/Relation/InternalDerived.hs

```diff
@@ -45,48 +48,167 @@ instance Ord a => Eq (ReflexiveRelation a) where
-    connect x y = ReflexiveRelation $ fromReflexive x `connect` fromReflexive y
-
-instance Ord a => Reflexive (ReflexiveRelation a)
-
--- TODO: Optimise the implementation by caching the results of symmetric closure.
 {-|  The 'SymmetricRelation' data type represents a /symmetric binary relation/
 over a set of elements. Symmetric relations satisfy all laws of the
-'Undirected' type class and, in particular, the
+'Undirected' type class and, in addition, the
 commutativity of connect:
 
 @'connect' x y == 'connect' y x@
 
 The 'Show' instance produces symmetrically closed expressions:
 
 @show (1     :: SymmetricRelation Int) == "vertex 1"
-show (1 * 2 :: SymmetricRelation Int) == "edges [(1,2),(2,1)]"@
+show (1 * 2 :: SymmetricRelation Int) == "edge 1 2"@
+
+The total order on graphs is defined using /size-lexicographic/ comparison:
+
+* Compare the number of vertices. In case of a tie, continue.
+* Compare the sets of vertices. In case of a tie, continue.
+* Compare the number of edges. In case of a tie, continue.
+* Compare the sets of edges.
+
+Here are a few examples:
+
+@'vertex' 1 < 'vertex' 2
+'vertex' 3 < 'Algebra.Graph.Relation.Symmetric.edge' 1 2
+'vertex' 1 < 'Algebra.Graph.Relation.Symmetric.edge' 1 1
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 = 'Algebra.Graph.Relation.Symmetric.edge' 2 1
+'Algebra.Graph.Relation.Symmetric.edge' 1 1 < 'Algebra.Graph.Relation.Symmetric.edge' 1 2
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 > 'Algebra.Graph.Relation.Symmetric.edge' 1 1 + 'Algebra.Graph.Relation.Symmetric.edge' 2 2
+'Algebra.Graph.Relation.Symmetric.edge' 1 2 < 'Algebra.Graph.Relation.Symmetric.edge' 1 3@
+
+Note that the resulting order refines the 'isSubgraphOf' relation and is
+compatible with 'overlay' and 'connect' operations:
+
+@'Algebra.Graph.Relation.Symmetric.isSubgraphOf' x y ==> x <= y@
+
+@'empty' <= x
+x     <= x + y
+x + y <= x * y@
 -}
 newtype SymmetricRelation a = SymmetricRelation { fromSymmetric :: Relation a }
-    deriving (Num, NFData)
+    deriving NFData
 
 instance Ord a => Eq (SymmetricRelation a) where
-    x == y = symmetricClosure (fromSymmetric x) == symmetricClosure (fromSymmetric y)
+    x == y = (fromSymmetric x) == (fromSymmetric y)
 
 instance (Ord a, Show a) => Show (SymmetricRelation a) where
-    show = show . symmetricClosure . fromSymmetric
-
--- TODO: To be derived automatically using GeneralizedNewtypeDeriving in GHC 8.2
-instance Ord a => Graph (SymmetricRelation a) where
-    type Vertex (SymmetricRelation a) = a
-    empty       = SymmetricRelation empty
-    vertex      = SymmetricRelation . vertex
-    overlay x y = SymmetricRelation $ fromSymmetric x `overlay` fromSymmetric y
-    connect x y = SymmetricRelation $ fromSymmetric x `connect` fromSymmetric y
-
-instance Ord a => Undirected (SymmetricRelation a)
+    show = show . filtered . fromSymmetric
+        where
+            filtered (Relation d r) = Relation d (Set.filter (uncurry (<=)) r)
```

Comment:
You perform this operation very frequently. I think it should be factored out into a separate internal function, which could be called `deduplicate`.